TogetherCrew: Community Health Check (#1482)
w3f/Grants-Program: applications/community-health-check.md

Our community health checks combine objective data (interactions between contributors) and subjective data (contributors perceptions about the community). Interactions between contributors are a vital data point: During interactions, members exchange knowledge, develop ideas, and influence each others perspective. Therefore, a community dashboard that measures interaction patterns, compared to solely activity, acknowledges that contributors do not act in isolation. Measuring perception data through pulse survey (tiny surveys with 3-5 questions) gives community leaders a quick and easily administered insights into how contributors feel.

Although our initial focus is community health metrics, the proposed approach sets the foundation for further applications. The combination of SNA and pulse surveys offers unprecedented actionable insights in real-time. Some of the potential applications and insights for the ecosystem are

> - Generate baseline metrics for community health / vibez to quantify and better understand the impact a specific event is having on a community and/or sub-groups within the community

Reviewer: > vibez

What do you mean by that?

Author: with vibez we mean the feeling/quality of the community

Reviewer: ah you mean _vibes_

> - Generate baseline metrics for community health / vibez to quantify and better understand the impact a specific event is having on a community and/or sub-groups within the community

Reviewer: ```suggestion
- Generate baseline metrics for community health / vibes to quantify and better understand the impact a specific event is having on a community and/or sub-groups within the community
```

- Identify measurements of decentralization to serve as KPIs or Insights metrics
- Build funnels to track member onboarding and identify areas for improvement
- Monitor specific topics like contributor wellbeing, alignment, community experience, etc. in near real-time
- Attract talent and investment with objective community health metrics instead of proxy metrics like member count or proposal count, or financial metrics such as TVL
- Help new contributors find context-rich mentors outside of the existing pool of well-known but time-poor candidates

### Rationale 

Social network data and its methodology (Social Network Analysis; SNA) is a structured way to visualize how communications, information, and value creation occur through an organization based on interaction graphs. Although a relatively new decision-making tool for leaders, SNA is gaining popularity over traditional survey tools. SNA has been linked to a wide range of insights, such as improving contributor retention, avoiding member burnout, predicting team performance and community resilience, identifying key contributors, enabling decentralization, and improving coordination.

> ![Data points for measuring community health](howmeasurecommhealth.jpg)

Reviewer: ```suggestion
![Data points for measuring community health](https://user-images.githubusercontent.com/5393704/224280603-03e8b7a7-4212-4a46-b957-ef06ebe3de5a.png)
```

Ideally, we avoid polluting the `applications` directory with images and reference them from an external URL instead.

### Ecosystem Fit

- Where and how does your project fit into the ecosystem?
Our project gives the ecosystem information about its health by measuring the interactions (conversations, relationships) between contributors. We consider this a "soft performance indicator", because it doesn't measure objective aspects (e.g., performance, dapps launched, bugs fixed) but centers on the humans of the ecosystem. 

- Who is your target audience (parachain/dapp/wallet/UI developers, designers, your own user base, some dapp's userbase, yourself)?
The community health check is for community managers and moderators to support them in their work by providing community metrics.

- What need(s) does your project meet?
Our project provides moderators, leaders, and contributors with a bird's eye view of what is happening inside the ecosystem. 

- Are there any other projects similar to yours in the Substrate / Polkadot / Kusama ecosystem?

> Our current approach rests on off-chain data. Other community analytics dashboard exists, fetching data from various off-chain data. However, they focus on posting behavior and not on the  interaction between members. Orbit.love reach metric, conceptually, comes close to our metrics, but technically also rests on individual posting behavior and not interaction data. 

Reviewer: > but technically also rests on individual posting behavior and not interaction data

Could you further elaborate on the difference between _individual posting behavior_ and _interaction data_? Given a network graph where the nodes represent the SN's participants and the edges between them their relationships (e.g. likes, comments, follows, etc.), does it mean that the further only focuses on the nodes while your solution would additionally take the edges into account?

Author: Yes you said it well. Individual posting behavior focuses only on the number of activities of a person, but not how these activities impact others. Edges are our interaction data (replies, mentions, emoji reactions)

Reviewer: Thanks. If that's the case, could you please incorporate this information into the doc?

#### Is this safe or legal/ Are you GDPR compliant?
Analyzing conversations in private channels (e.g., Discord) is within a grey zone of private and public conversations and a sensitive issue. Servers and channels that are not gated in anyway (token or role) are akin to public conversations as there are no barriers for people to enter and join the conversations. On the other hand, once users need to fulfil certain criterias or pay in order to access a server/channel, the conversation could be perceived as private. This distinction has ethical and legal consequences. 

For this reason, we have worked with a legal counsel via LexDAO. As part of this, two lawyers have reviewed the process from fetching data and presenting results to ensure we are within GDPR laws. On Wednesday 22nd February we had a GDPR session for our team and  implemented the lawyers' recommendations.
- anti-virus software to be installed on the personal laptops of the two data scientists who have access to the raw data
- a data transfer agreement will be signed with the data scientist residing in the US

> - We will prepare a record of processind activities

Reviewer: ```suggestion
- We will prepare a record of processing activities
```

Early development of this project has been funded by [Aragon](https://forum.aragon.org/t/real-time-community-health-analytics), Polygon DAO, [Aave](https://noturhandle.gitbook.io/rndao/proposals/proposals/grant-proposal-aave), [MetaCartel](https://forum.metacartel.org/t/grant-request-community-health-a-project-by-rndao/2354) and Near.

## Development Roadmap :nut_and_bolt:

### Overview

- **Total Estimated Duration:**  10 weeks
- **Full-Time Equivalent (FTE):**  6
- **Total Costs:**  15 000 USDC

We are following a grant-matching processes to spread the risk among different communities. 

### Milestone 1 - Defining metrics

- **Estimated duration:** 4 weeks
- **FTE:**  1
- **Costs:** 5000 USDC

Research social network metrics that are viable to measure the health of a community on Twitter, and how they integrate with our existing community health framework. This will build on our existing framework, extending it to metrics that have been tested and used in Twitter communities. For example, we will look into approaches to build the network, virality and clustering/fragmentation of very large online communities. 

> Note: We have done an evaluation to decide wheter Reddit or Twitter would be a better option. Given other requests we have received, we have decided that building an integration and health metrics for Twitter is a more viable path for us.

Reviewer: > Given other requests we have received, we have decided that building an integration and health metrics for Twitter is a more viable path for us.

I'm curious to learn what led to the decision to pick Twitter over Reddit.

Reviewer: @katerinabc ping

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |

> | 0a. | License | The results will be published open-access (no paywall or copyright). |

Reviewer: Could you specify the license you're planning to use?

Reviewer: @katerinabc this is still not specified. Please choose one of the [supported licenses](https://github.com/w3f/Grants-Program/#guidelines):

> We prefer Apache 2.0, but GPLv3, MIT or Unlicense are also acceptable.

If you're undecided, you can specify multiple ones and add a note, that you will pick one of them

Author: Apologies. I worked for a long time in academia and there it's just a given that every written document can be re-used by others just with referencing. I'm gonna add multiple ones and decide later.

| 0b. | Documentation | No documentation necessary.|
| 0c. | Testing and Testing Guide | Testing and Testing Guide not necessary. |

> | 0d. | Docker | A docker file is not necessary for using the community health check. |

Reviewer: You can remove these if you're not planning to deliver anything.

Author: Thanks. I'll  leave them in so that it is clear that this is not an error of omission.

Reviewer: @katerinabc see my latest suggestion on that block.

> | 0e. | Article| We will publish an article (technical document) describing the metrics, the insights (so-what), and limitations. This article will also explain how the Twitter community is build. |

Reviewer: Could you add more details here?

Author: I've added more information. If you are looking for something else, let me know.

Reviewer: Thanks. Could you please add more info on which relationships between the nodes (i.e. edges) you're planning to analyse? Ideally, provide a full list of supported edges so we can easily verify the milestone, once delivered.

Author: Please have a look at milestone 1. I have specified how we are building the network by adding the following information:

We will build a directed network, where nodes are always Twitter users. From a network assembly perspective, we will not differentiate between accounts representing people and those representing communities or organizations. The edges between an user profiles are either a reply, quote, mention, retweets, or likes. Thus, a tie from user A to user B exists if (1) user A replies to user B, user A quotes user B, user A mentions user B, user A retweets user B, or user A likes user B's tweet. At this moment, we will created weighted edges, not making a conceptual difference between the interaction type (reply, quote, mention, retweets and likes).

### Milestone 2 — Twitter community health dashboard

- **Estimated duration:** 6 weeks
- **FTE:**  5
- **Costs:** 10 000 USDC

This milestone implements the work from the previous milestone by building the dashboard.

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |

> | 0a. | License | The code for data fetching and storage aisre already available in a public github folder (Unlicensed). We will specify an OSS license. |

Reviewer: ```suggestion
| 0a. | License | The code for data fetching and storage is already available in a public GitHub folder (Unlicensed). We will specify an OSS license. |
```

Could you add a link to the GitHub repo since it's already available?

| 0b. | Documentation | We will provide a tutorial for users to understand how to run the health checks themselves. Our current version will be updated to include recent development.|

> | 0c. | Testing and Testing Guide | Running the data pipeline and analyzing the data will be covered by tests to ensure functionality. We will describe how to run the tests|

Reviewer: ```suggestion
| 0c. | Testing and Testing Guide | Running the data pipeline and analyzing the data will be covered by tests to ensure functionality. We will describe how to run the tests|.
```

> | 0d. | Docker | We will run a docker file to tests the functionality. |

Reviewer: ```suggestion
| 0d. | Docker | We will deliver a docker file to tests the functionality. |
```

| 0e. | Article | In addition to the dashboard, we write a handout. This is a walkthrough of the dashboard, explaining each metric, if the score is good or bad, and a list of recommendations.  |

> | 1. | Twitter data integration | We will create a data pipeline fetching data from Twitter using their API. The user will have several options to define their community (e.g., everyone who follows the official Polkadot/Kusama account, everyone who follows the account and interacts with the content). |

Reviewer: - Will you write the integration for a paid version of the API or the free tier?
- Can you elaborate on the throughput that this your crawler will be able to deliver (e.g. how many requests/min, how many tweets, comments, likes, etc. can be retrieved by some unit of time).
- How about other accounts, e.g. parachains, dApp developers and other teams that build on Substrate?

Author: The updated proposal will have more information.

Reviewer: Similarly to the research deliverable in M1, can you please add more specific information about which edges will be supported?

Author: I have decided to not repeat the same information here than for Milestone 1. Milestone 1 is our research phase, while milestone 2 is the implementation. Hence, decisions we are making during milestone 1 will impact milestone 2.

Reviewer: @katerinabc lovely, thanks for adding that info.

### Milestone 1 - Defining metrics

- **Estimated duration:** 4 weeks
- **FTE:**  1
- **Costs:** 5000 USDC

Research social network metrics that are viable to measure the health of a community on Twitter, and how they integrate with our existing community health framework. This will build on our existing framework, extending it to metrics that have been tested and used in Twitter communities. For example, we will look into approaches to build the network, virality and clustering/fragmentation of very large online communities. 

Note: We have done an evaluation to decide wheter Reddit or Twitter would be a better option. In the discussion we had with other communities, we were more often asked about an integration with Twitter than with Reddit. Hence, from a scaling perspective, we decided that focusing first on Twitter makes more sense with us.

That being said, we realize that Twitter is undergoing a lot of changes, and we might have to do a last minute pivot to another platform. We are building our analytical scripts in such a way that they can easily be used for other platforms. Therefore, the only thing that has to be changed in the milestones is the name of the platform. 

> 
> 
> | Number | Deliverable | Specification |
> | -----: | ----------- | ------------- |
> | 0a. | License | The results will be published open-access (no paywall or copyright). |
> | 0b. | Documentation | No documentation necessary.|
> | 0c. | Testing and Testing Guide | Testing and Testing Guide not necessary. |
> | 0d. | Docker | A docker file is not necessary for using the community health check. |
> | 0e. | Article| We will publish an article (technical document) describing the metrics, the insights (so-what), and limitations. This article will also explain how the Twitter community is build (e.g., who are nodes, when there is an edge between two people, who is excluded/included and why). The article will not be behind a paywall. The article will be written for an audience comfortable with data analysis. |

Reviewer: ```suggestion

Milestone 1 will be focused on research. Hence, we're not going to deliver the following usually mandatory deliverables:

- Documentation
- Testing and Testing Guide
- Docker

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |
| 0a. | License | The results will be published open-access (no paywall or copyright). |
| 0b. | Article | We will publish an article (technical document) describing the metrics, insights (so-what), and limitations. This article will also explain how the Twitter community is built (e.g., who are nodes, when there is an edge between two people, who is excluded/included, and why). The article will not be behind a paywall. The article will be written for an audience comfortable with data analysis. |
```

Reviewer: I still think it's easier to skip the deliverables that you're not delivering. This way, it will also be easier to validate the results.

### Milestone 2 — Twitter community health dashboard

- **Estimated duration:** 6 weeks
- **FTE:**  5
- **Costs:** 10 000 USDC

This milestone implements the work from the previous milestone by building the dashboard. It will be build using the Twitter API v2. This comes with the following rate limits: 
- 900 tweet lookups/user/15-minute time window
- 900 user timeline lookups/user/15-minute time window
When querying a tweet, we will get information about the number of comments and likes 

Given the rate limit, we will see how users will be able to combine different accounts (e.g., parachains, dApp developers and other teams that build on Substrate) to create a holistic picture of their community. 

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |

> | 0a. | License | Apache 2.0 |

Reviewer 2: Here is where I'm suggesting to update with the full Apache 2.0 text instead of just the header. Thanks!
